// @aws-cdk/cx-api/lib/cloudformation-artifact.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const path = require("path");
const assets_1 = require("./assets");
const cloud_artifact_1 = require("./cloud-artifact");
class CloudFormationStackArtifact extends cloud_artifact_1.CloudArtifact {
    constructor(assembly, name, artifact) {
        super(assembly, name, artifact);
        if (!artifact.properties || !artifact.properties.templateFile) {
            throw new Error(`Invalid CloudFormation stack artifact. Missing "templateFile" property in cloud assembly manifest`);
        }
        const properties = (this.manifest.properties || {});
        this.templateFile = properties.templateFile;
        this.parameters = properties.parameters || {};
        this.name = this.originalName = name;
        this.template = JSON.parse(fs.readFileSync(path.join(this.assembly.directory, this.templateFile), 'utf-8'));
        this.assets = this.findMetadataByType(assets_1.ASSET_METADATA).map(e => e.data);
    }
}
exports.CloudFormationStackArtifact = CloudFormationStackArtifact;
